Initial set of changes for ui correctness state when saving a problem

staubina opened the pull request:
TNL-1955
## [TNL-1955](https://openedx.atlassian.net/browse/TNL-1955)

### Description

Updates so that when an answer is submitted, changed, and then saved the ui is cleared.

### Sandbox
- [x] Build a sandbox for your branch and add a link here

https://staubina-save.sandbox.edx.org

### Testing
- [x] i18n
- [x] RTL
- [x] safecommit violation code review process
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @cahrens 
- [x] Code review: @muhammad-ammar 
- [x] Doc Review: @catong , FYI I am not sure if this requires doc updates.
- [x] Product review: @sstack22 

### Post-review
- [x] Rebase and squash commits

staubina commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -789,7 +793,9 @@ def _extract_html(self, problemtree):  # private
             if problemid in self.correct_map:
                 pid = input_id
-                status = self.correct_map.get_correctness(pid)
+                if not self.saved_answers:
```
Note: This was tested around correctness.  This will not impact the correctness state in the correct map and will respect the current score.  It will return an unanswered state when an answer has been saved since last submit.

cahrens replied:
I wonder if here you should show the correctness if the saved answer matches the submitted answer.

staubina commented:
@cahrens Who should I add as second dev reviewer?

staubina commented:
@muzaffaryousaf Is someone available on your team to review this tomorrow?

cahrens commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -143,6 +143,7 @@ def __init__(self, problem_text, id, capa_system, capa_module,  # pylint: disabl
             state (dict): containing the following keys:
                 - `seed` (int) random number generator seed
                 - `student_answers` (dict) maps input id to the stored answer for that input
+                - 'saved_answers' (Boolean) True is answer has been saved since last submit.
```
I would expect "saved_answers" to be a dict (like "student_answers"). How about has_saved_answers?

Note also the typo "True is answer".

cahrens commented:
@sstack22 @staubina do we care about the case where the saved answer matches the last submitted answer? For example,

1) Answer question incorrectly and press Submit
2) Save the exact same answer
3) Refresh the page-- no correctness is shown

Same is true if at step #1 you answer (and submit) correctly

cahrens commented:
@sstack22 @staubina Another product question-- after I have clicked Submit and then select a different answer, should the correctness marks go away? Depending on the problem type, they don't go away until refresh. https://openedx.atlassian.net/browse/TNL-2404 indicates that this may be a change in behavior from some earlier state.

cahrens commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -19,19 +19,23 @@
-                    if status == 'correct':
+                   <%
+                    if saved_answers:
+                        correctness = status.classname
```
What is status.classname in this case?

staubina replied:
unanswered

cahrens replied:
Why is status just a string in the other cases below? Ideally status.classname would suffice for all.

cahrens commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -162,6 +162,8 @@ class CapaFields(object):
                        scope=Scope.user_state, default={})
     input_state = Dict(help=_("Dictionary for maintaining the state of inputtypes"), scope=Scope.user_state)
     student_answers = Dict(help=_("Dictionary with the current student responses"), scope=Scope.user_state)
+    saved_answers = Boolean(help=_("Whether or not the answers have been saved since last submit"),
```
Where is the saved answer actually stored? Why is it necessary to have an additional boolean instead of looking to see if a saved answer exists?

staubina replied:
student_answers, I will look to see if it can be cleared on submit.

staubina commented:
@cahrens It goes away for some problem types and not others.  I can get the types that go away.  We only detect answers on certain problem types right now.

JAAkana commented:
@sstack22 Thanks for asking me to take a look as a user! 
Agreed that this resolves the issue in TNL-1955 - students will no longer see feedback that doesn't apply to their current selection.

Taking a step back, I still find the experience confusing. After doing the following steps in a multiple choice question:
1. Submit a correct answer.
2. Select a different answer, Save.
3. Refresh the page.
I look at the page, and can see: my point total, a selected option, and no feedback.

It's not clear to me how a student can understand how their point value relates to the selection.

In my shoot-the-moon, if-i-had a pony world, students would be able to parse the feedback, 'You submitted <x> and received 1 point (correct); You've saved response <y> but not submitted it.' Or alternatively, the "Your answers have been saved but not graded. Click 'Submit' to grade them" message sticking around after page refresh.

muzaffaryousaf commented:
@muhammad-ammar Can you review this PR, if you have bandwidth today or tomorrow ?

muhammad-ammar commented:
@muzaffaryousaf I can look at it tomorrow.

staubina commented:
@JAAkana This was a concern I also had discussed this briefly with @sstack22, It should be possible to persist the save notification until a submit occurs.

staubina commented:
jenkins run lettuce

staubina commented:
jenkins run lettuce

sstack22 commented:
@staubina, @JAAkana - let's look into the save notification persisting until submit.

staubina commented:
@cahrens @sstack22 @muhammad-ammar, The Sandbox is updated.  I still need to look into the changes @sstack22 and @JAAkana have requested to the save notification.

muhammad-ammar commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -789,7 +793,9 @@ def _extract_html(self, problemtree):  # private
                 pid = input_id
-                status = self.correct_map.get_correctness(pid)
+                if not self.has_saved_answers:
+                    status = self.correct_map.get_correctness(pid)
```
@staubina Can you add a comment for this condition to help understand what we are doing here?

muhammad-ammar commented:
@staubina Can you point me to a test which verifies that below steps gives us the expected behaviour? 

1. Attempt answer with a wrong choice.
2. Select another choice and click save
3. Reload and it shouldn't show the cross sign

staubina commented:
I am making the requested changes.  I will send out a notification when I am done and the Sandbox is updated.

staubina commented:
jenkins run bokchoy

staubina commented:
@muhammad-ammar @cahrens I have addressed your comments and updated the PR. The Sandbox is also updated.

@muhammad-ammar You will also see that I have added a test for the scenario you mentioned.

muhammad-ammar commented on common/test/acceptance/tests/lms/test_lms_problems.py:
```diff
@@ -225,6 +225,51 @@ def test_notification_updates(self):
+    def test_status_removed_after_save_before_submit(self):
+        """
+        Verifies that the status on a problem is removed after a save and before a second submission
+        """
```
@staubina Please add BDD style docstring. You can see an example [here](https://github.com/edx/edx-platform/blob/master/common/test/acceptance/tests/video/test_video_module.py#L468).

muhammad-ammar commented:
> For a problem with hints and feedback(for example Multiple Choice with Hints and Feedback) . Incorrect feedback is shown after we select the correct option and save and then reload the page. Here are the steps

> - Create Multiple Choice with Hints and Feedback problem
> - Select and submit the wrong answer. You will see the feedback for incorrect option
> - Select the correct answer and save
> - Reload the page
> - You will see that correct option is selected but feedback for incorrect option is shown there
> 
> I think this behaviour needs to be changed.
> 

@staubina I am still seeing the above behaviour as i mentioned before. So this behaviour will remain as it is?

<img width="1042" alt="screen shot 2016-11-15 at 6 44 47 pm" src="https://cloud.githubusercontent.com/assets/6767924/20308008/ef56ee74-ab63-11e6-9499-574a25c86024.png">

Also the new save notification as seen the above screenshot is wrapped although we have free space? Is this intended?

staubina commented:
@muhammad-ammar I am not seeing that behavior.  I will try again though and try to recreate.

staubina commented:
@muhammad-ammar Ahhh.  The Feedback on the bottom.  Let me look into that now.

staubina commented:
> Also the new save notification as seen the above screenshot is wrapped although we have free space? Is this intended?

This is due to the styling of the Notification message and is a result of the hidden review button and the length of the button when translated.  @sstack22 or @chris-mike Are you concerned with the positioning of this message?  We can alter the style if needed. I have not changed it in this PR.

cahrens commented on common/lib/capa/capa/templates/textline.html:
```diff
@@ -7,7 +7,6 @@
     <div class="text-input-dynamath_data ${doinline}" data-preprocessor="${preprocessor['class_name']}"/>
     <div class="script_placeholder" data-src="${preprocessor['script_src']}"/>
 % endif
-
```
Very minor nit: you could remove the edits to this file from your PR so isn't linked (since it had no actual code changes).

cahrens commented:
@staubina I looked through the code, and it looks great! Also played a bit with the sandbox. Let me know when the final commit is up.

staubina commented:
@muhammad-ammar @cahrens 

The PR is updated with the requested changes.  Also, the Sandbox is updated as well.

I am looking at some safecommit violations to at least clean up some of them before merging.

cahrens commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -61,7 +61,7 @@
     <div class="capa_alert">${submitted_message}</div>
     %endif
-    % if msg:
+    % if msg and status.classname != 'unanswered':
```
@staubina the span for "message" is included in every problem type HTML file, I believe. Why did you only change these 2?

Also, are we sure we never want to show a message if the problem is unanswered? I feel like I want to look more at what message can be.

cahrens commented on lms/templates/problem.html:
```diff
@@ -98,6 +101,7 @@ <h3 class="hd hd-2 problem-header" id="${ short_id }-problem-title" aria-describ
       notification_name='save',
-      notification_message=''"
+      notification_message=save_message,
+      is_hidden=not has_saved_answers"
```
Instead of adding this "is_hidden" property, can't you key off of whether or not a non-empty save_message exists? Then the server code (vs. template) is deciding what to render based on state.

cahrens commented:
👍 after you fix the one issue I mentioned in choicetext.html

staubina commented:
jankins run bokchoy

staubina commented:
jenkins run bokchoy

staubina commented:
@catong Please take a quick look at this and let me know if you think I should change the Save text or anything else.

staubina commented:
@sstack22 and @muhammad-ammar Please take anther look if you have a moment today.  I would like to merge this.

muhammad-ammar commented:
@staubina great work 👍

staubina commented:
@cahrens Please take one quick look at my most recent commit and let me know if you are ok with the change.

staubina commented:
I am updating the Sandbox now.

cahrens commented:
@staubina I'm fine with the change, but with no test this is very likely to break in the future (simple class name change in the HTML toasts it). Please look to see what test coverage exists already of this message area, and then see how much effort it would take to extend the coverage for this new case.

staubina commented:
@cahrens I did not see any existing tests, so I added one that check to see the message and then checks that it is hidden.

staubina commented:
@sstack22 The Sandbox is updated with the changed to feedback

cahrens commented:
👍

sstack22 commented:
👍
